### PR TITLE
add flexible answers component

### PR DESCRIPTION
**Pull request by jhackett1**
this adds a flexible answers component to enable the app to display records from the [care recording app](https://github.com/LBHackney-IT/social-care-recording)

here's a working demo of the component: https://codepen.io/jhackett1/pen/KKWwoJy?editors=0010

because this component, by its nature, decodes a block of json which could have wildly different nested structures, the typescript interfaces look a bit unusual - i've been as specific with them as is safe.

it adds the component to `components/Cases/CaseRecap/CaseNote`, where it is conditionally rendered if the record has the `form_name` of "Case recording".

this is far from ideal and should be replaced with a more robust means of telling what "kind" of recording it is.

### CODE REVIEW

File: components/Cases/CaseRecap/CaseNote.tsx [hunk 1]
```diff
@@ -13,6 +14,11 @@ interface Props {
 const CaseNote = ({ personId, recordId }: Props): React.ReactElement => {
   const { data: record, error: recordError } = useCase(recordId);
 
+  if (record?.caseFormData?.case_note_description) {
```
Comment (JohnFarrellDev) [inline components/Cases/CaseRecap/CaseNote.tsx]: Do we need to test this?

What is case_note_description is an empty string, malformed JSON, JSON in a format we expect.

File: components/FlexibleAnswers/FlexibleAnswers.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  </ul>
+);
+
+const RepeaterGroupAnswers = ({
+  data,
+}: {
+  data: (string | RepeaterGroupAnswer)[];
+}): React.ReactElement => (
+  <ul className="govuk-list lbh-list">
+    {data.map((item, i) => (
```
Comment (JohnFarrellDev) [inline components/FlexibleAnswers/FlexibleAnswers.tsx]: If this is a repeater group can new items be added, can items be removed?

If so we need to be careful with using the index as a key.

However do we have anything appropriate here that was could use as a unique identifier for each element.

File: tsconfig.json [hunk 3]
```diff
@@ -17,6 +17,6 @@
     "jsx": "preserve",
     "baseUrl": "."
   },
-  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
+  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", "components/FlexibleAnswers/blah"],
```
Comment (JohnFarrellDev) [inline tsconfig.json:20]: What is this change for?

File: components/FlexibleAnswers/FlexibleAnswers.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      // questions and answers
+      [key: string]: string | (string | RepeaterGroupAnswer)[];
+    };
+  };
+}
+
+const FlexibleAnswers = ({ data }: Props): React.ReactElement => {
+  return (
+    <div>
+      {Object.keys(data).map((stepName) => (
```
Comment (JohnFarrellDev) [inline components/FlexibleAnswers/FlexibleAnswers.tsx:54]: Object.entries here could be nice as we are having to access the value below.

File: components/FlexibleAnswers/FlexibleAnswers.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        <section key={stepName}>
+          <h3 className="lbh-heading-h3 govuk-!-margin-top-6 govuk-!-margin-bottom-4">
+            {stepName}
+          </h3>
+          <dl className="govuk-summary-list lbh-summary-list">
+            {Object.keys(data[stepName]).map((questionName) => (
```
Comment (JohnFarrellDev) [inline components/FlexibleAnswers/FlexibleAnswers.tsx:60]: I think the map inside of the map makes this confusing too read.

It also means the typeof check can be moved out of the JSX and then we could avoid the cast I believe.

**Review by JohnFarrellDev — COMMENTED** [review]
Small changes, looks good

File: tsconfig.json (hunk 3 above)
Comment (jhackett1) [reply tsconfig.json:20]: what the heck is that. will revert

File: components/Cases/CaseRecap/CaseNote.tsx (hunk 1 above)
Comment (jhackett1) [reply components/Cases/CaseRecap/CaseNote.tsx]: not happy with this whole bit of the code. we need a more foolproof way to know when to render flexible answers i think

File: components/FlexibleAnswers/FlexibleAnswers.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+          : data[questionName]}
+      </li>
+    ))}
+  </ul>
+);
+
+const RepeaterGroupAnswers = ({
+  data,
+}: {
+  data: (string | RepeaterGroupAnswer)[];
```
Comment (JohnFarrellDev) [inline components/FlexibleAnswers/FlexibleAnswers.tsx:25]: We're using `data` a lot as a variable name, could we use something more descriptive?

Comment (jhackett1) [reply components/FlexibleAnswers/FlexibleAnswers.tsx:25]: answers?

File: components/Cases/CaseRecap/CaseNote.tsx [hunk 7]
```diff
@@ -23,6 +25,19 @@ const CaseNote = ({ personId, recordId }: Props): React.ReactElement => {
   if (!record || !fileData) {
     return <Spinner />;
   }
+
+  // this needs to be fixed with a better way of distinguishing
+  if (
+    record?.caseFormData?.case_note_description &&
+    record.caseFormData.form_name === 'Case recording'
```
Comment (jhackett1) [inline components/Cases/CaseRecap/CaseNote.tsx:32]: how might we better distinguish whether to render flexible answers?

**Review by anabebs — APPROVED** [review]
all good

**Pull request merged**